alphagov/finder-frontend
# Add GA4 finder "change" tracker

Hi @andysellick / @JamesCGDS - would you be able to review this when you get a chance? I've tried to explain it as best as I could, but let me know if something doesn't make sense. Thanks. 

Related to https://github.com/alphagov/govuk_publishing_components/pull/3362


## What
- Adds tracking to the `/search` page using `ga4-form-tracker`
- Adds a new `Ga4FinderTracker` which tracks whenever our GOVUK search or Finders are changed.
- Adds `data-` attributes that our `govuk_publishing_components` `Ga4FinderTracker` makes use of.


Related trello cards: https://trello.com/c/zCIwJ8pP/515-add-tracking-search-box-on-search-finder-pages https://trello.com/c/aEXGK5cE/527-add-tracking-select-filter-on-search-finders-selectcontent
https://trello.com/c/sH9vKqA4/528-add-tracking-remove-filter-on-search-finders-selectcontent
https://trello.com/c/ZQeyh2qb/530-add-tracking-sort-by-search-finders-selectcontent

## Why
- The new "finder tracker" is needed because the finders are updated with client side JavaScript. They fire a `change` event whenever a keyword/filter is changed. Therefore we need to hook into this to track it in GA4.
- As each finder is one big form, it fires one big change event if anything in the search changes. Therefore we need a way to know which element triggered the change, so that we can send the correct GA4 event. We use `event.target` as well as `data-` attributes on the input or input parent, which gives us a simple way to categorise the change and build our GA4 object.
- The code was originally in the other PR: https://github.com/alphagov/govuk_publishing_components/pull/3362 but now it is in this repo.

## Brief explanation of how it works

1. In finder-frontend, we define some data-attributes on the finder `<form>` elements. 
    a. Firstly, we put a `data-ga4-filter-container` on the element that wraps all the filters. This will be needed to set `index_section` on each filter.
    b. Next, we add a `data-ga4-section` on each filter with the `data-ga4-filter-container` element. This is needed to identify where we need to add `data-ga4-index` for our `index_section`s. This key is also used to identify the `section` value to add to the GTM object. E.g. for a Topics filter, the attribute would be `data-ga4-section="Topics"`
   c. Next, on each element that can cause a `change` event on the `<form>` we add an attribute called `data-ga4-change-category` which gives us 1. metadata on what that element changes about the search results and 2. the HTML type that the element is. For example, on a `<select>` element that changes the filters, we would add `data-ga4-change-category="update-filter select"`. On the search box we would add `data-ga4-change-category="update-keyword text".

2. On the loading of a `finder`, we call `window.GOVUK.analyticsGa4.Ga4FinderTracker.setFilterIndexes()` in `finder-frontend`. This will grab the `data-ga4-filter-container`, look for any `data-ga4-section` elements inside it, and set an `index_section` on each section it finds.
3. In `finder-frontend`, when the search results are changed, we pass through the `event.target` and `data-ga4-change-category` to `Ga4FinderTracker.trackFormChange()`.
4. `trackFormChange()` will grab the "value" of the Filter that caused the change. For example the name of the checkbox, or the new keyword that was just searched for.
5. If it was a filter that caused the `change` event, then `trackFormChange()` will try and determine wether the filter that changed was a `Add filter` or `Remove filter` event.
6. It will then use a `switch` statement to build the GTM object based on which `data-ga4-change-category` was passed. 

I've also written a markdown documentation file. Hope how I've built this makes sense.

## How to test this
1. Run `finder-frontend` with this branch checked out
2. Go to http://127.0.0.1:3062/search/all and change something in the search (e.g. a keyword, add a filter, remove a filter, or click "Clear all filters" in mobile view.) Check the `dataLayer` for an `event_data` event (it's usually index 7 in the dataLayer). Also check the finders without any filters e.g. http://127.0.0.1:3062/government/people and ones with radio buttons e.g. http://127.0.0.1:3062/official-documents .

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.
